### PR TITLE
Implementation UI warnings for outdated shipped feature

### PR DESCRIPTION
**Pull request by KyleJu**
A follow-up for #4481. Implement warning banners for outdated shipped features. If a feature is not scheduled to ship in the upcoming two milestones, it will attempt to find the `latest shipping milestone <= latest stable milestone`, as its `closestShippingDate`. If it exists, a feature is considered shipped.

A shipped feature is outdated `if accurate_as_of < closestShippingDate`. If outdated, it will show:
- a warning for feature authors if the viewer is an author, or
- a warning for a regular reader `if accurate_as_of + 9 weeks < current date` ( last updated 2~ months ago)

A feature can only be considered as an upcoming feature or shipped feature, but not both.

For authors:
Screenshot 2024-10-30 11 54 18 PM

For all readers:
Screenshot 2024-10-30 11 55 34 PM

### CODE REVIEW

**Comment by KyleJu** [comment]
@jyasskin Still working through a few tests but implementation is complete, based on [your comment](https://github.com/GoogleChrome/chromium-dashboard/pull/4481#discussion_r1821224129). Feel free to take a look!

File: client-src/elements/chromedash-feature-page.ts [hunk 1]
```diff
@@ -128,11 +130,31 @@ export class ChromedashFeaturePage extends LitElement {
...
+      .then(([newMilestonesInfo]) => {
+        this.closestShippingDate = newMilestonesInfo[milestone]?.final_beta;
+        this.hasShipped = true;
+      })
+      .catch(() => {
+        showToastMessage(
+          'Some errors occurred. Please refresh the page or try again later.'
+        );
+      });
+  }
```
Comment (jyasskin) [inline client-src/elements/chromedash-feature-page.ts:151]: Any reason not to write this as an async function?

```suggestion
  async fetchClosestShippingDate(milestone: number) {
    if (milestone === 0) {
      return;
    }
    try {
      const newMilestonesInfo = await window.csClient.getSpecifiedChannels(
        milestone,
        milestone
      );
      this.closestShippingDate = newMilestonesInfo[milestone]?.final_beta;
      this.hasShipped = true;
    } catch {
      showToastMessage(
        'Some errors occurred. Please refresh the page or try again later.'
      );
    }
  }
```

You could even just return the closestShippingDate and let the caller handle assigning it into member variables.

File: client-src/elements/chromedash-feature-page.ts [hunk 2]
```diff
@@ -164,13 +186,72 @@ export class ChromedashFeaturePage extends LitElement {
...
+      } else {
+        this.fetchClosestShippingDate(latestMilestone);
+      }
     }
   }
 
+  /**
+   * Determine if it should show warnings to a feature author, if
+   * a shipped feature is outdated, and it has edit access.*/
+  isWarningShippedFeatureOutdatedForAuthor() {
```
Comment (jyasskin) [inline client-src/elements/chromedash-feature-page.ts:210]: Maybe
```suggestion
  shouldWarnThatShippedFeatureOutdatedForAuthor() {
```
or
```suggestion
  isShippedFeatureOutdatedForAuthor() {
```

File: client-src/elements/utils.ts [hunk 3]
```diff
@@ -663,17 +663,20 @@ export function extensionMilestoneIsValid(value, currentMilestone) {
  *
  *  @param accurateAsOf The accurate_as_of date as an ISO string.
  *  @param currentDate The current date in milliseconds.
+ *  @param gracePeriod The grace period in milliseconds. Defatuls
```
Comment (jyasskin) [inline client-src/elements/utils.ts:666]:
```suggestion
 *  @param gracePeriod The grace period in milliseconds. Defaults
```

**Review by jyasskin — COMMENTED** [review]
The UI looks good to me. Thanks!

File: client-src/elements/chromedash-feature-page.ts (hunk 1 above)
Comment (KyleJu) [reply client-src/elements/chromedash-feature-page.ts:151]: Done!

File: client-src/elements/utils.ts (hunk 3 above)
Comment (KyleJu) [reply client-src/elements/utils.ts:666]: oops Thanks!

File: client-src/elements/chromedash-feature-page.ts [hunk 4]
```diff
@@ -164,13 +184,74 @@ export class ChromedashFeaturePage extends LitElement {
...
+  /**
+   * A feature is outdated if it has shipped, and its
+   * accurate_as_of is before its latest shipping date before today.*/
+  isShippedFeatureOutdated(): boolean {
+    // Check if a feature has shipped.
+    if (!this.hasShipped) {
+      return false;
+    }
+
+    if (!this.feature.accurate_as_of) {
```
Comment (jrobbins) [inline client-src/elements/chromedash-feature-page.ts:241]: Should this case return false?

We set accurate_as_of when a feature is created.  The code to do that has been live for about two years.  Anything that still lacks accurate_as_of is probably an old entry.

If this returns true, then lots of old entries will start to show this banner.  I doubt that the feature owners for those old features will come back to verify and set accurate_as_of.

Comment (KyleJu) [reply client-src/elements/chromedash-feature-page.ts:241]: I see. That is a fair point! I will treat it as not outdated and will leave a comment inline

**Comment by KyleJu** [comment]
@jyasskin I will go ahead and merge this so it can be released this week. Feel free to raise any issues!

**Pull request merged**